connectors/src/connectors/github/lib/github_api.ts

```diff
@@ -526,3 +535,244 @@ async function getOctokit(installationId: string): Promise<Octokit> {
     },
   });
 }
+
+// Repository processing
+
+const asyncPipeline = promisify(pipeline);
+
+const EXTENSION_WHITELIST = [
+  ".js",
+  ".ts",
+  ".tsx",
+  ".jsx",
+  ".rb",
+  ".py",
+  ".rs",
+  ".go",
+  ".swift",
+  ".css",
+  ".html",
+  ".less",
+  ".sass",
+  ".scss",
+  ".php",
+  ".java",
+  ".yaml",
+  ".yml",
+  ".md",
+  ".c",
+  ".h",
+  ".cc",
+  ".cpp",
+  ".hpp",
+];
+
+const FILENAME_WHITELIST = [
+  "README",
+  "Dockerfile",
+  "package.json",
+  "Cargo.toml",
+];
+
+const DIRECTORY_BLACKLIST = [
+  "node_modules",
+  "vendor",
+  "dist",
+  "build",
+  "coverage",
+  "pkg",
+  "bundle",
+  "built",
+  "eggs",
+  "downloads",
+  "env",
+  "venv",
+  "tmp",
+  "temp",
+  "debug",
+  "target",
+];
+
+async function* getFiles(dir: string): AsyncGenerator<string> {
+  const dirents = await readdir(dir, { withFileTypes: true });
+  for (const dirent of dirents) {
+    const res = resolve(dir, dirent.name);
+    if (dirent.isDirectory()) {
+      // blacklist
+      if (DIRECTORY_BLACKLIST.includes(dirent.name)) {
+        continue;
+      }
+      yield* getFiles(res);
+    } else {
+      yield res;
+    }
+  }
+}
+
+// This function returns file and directories object with parent, internalIds, and sourceUrl
+// information. The root of the directory is considered the null parent (and will have to be
+// stitched by the activity).
+export async function processRepository(
+  installationId: string,
+  login: string,
+  repoName: string,
+  repoId: string
+) {
+  const octokit = await getOctokit(installationId);
+
+  const { data } = await octokit.rest.repos.get({
+    owner: login,
+    repo: repoName,
+  });
+  const defaultBranch = data.default_branch;
+
+  let { data: tarballStream } = await octokit.request(
+    "GET /repos/{owner}/{repo}/tarball/{ref}",
+    {
+      owner: login,
+      repo: repoName,
+      ref: defaultBranch,
+    }
+  );
+
+  // Create a temp directory.
+  const tempDir = await mkdtemp(join(tmpdir(), "repo-"));
+  const tarPath = resolve(tempDir, "repo.tar.gz");
+
+  // Convert ArrayBuffer to stream if necessary
+  if (tarballStream instanceof ArrayBuffer) {
+    // Wrap ArrayBuffer with a stream
+    const stream = new Readable();
+    stream.push(Buffer.from(tarballStream));
+    stream.push(null); // Signal the end of the stream
+    tarballStream = stream;
+  }
+
+  // Save the tarball to the temp directory.
+  await asyncPipeline(tarballStream as Readable, createWriteStream(tarPath));
```

Comment:
`nit` pipeline are useful when there is an operation in between read and write. Here a simple `pipe` could do the trick:

Something like:
```
const fileStream = createWriteStream(tarPath);
request(<gh_link>).pipe(fileStream);
```